numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
```

Comment:
Being the one that first suggested letting `np.array([])` through, I now think that Sebastian's consistency argument is more convincing. I think it is best that this PR follows what indexing does, gets merged quick, and if you feel strongly about this, we can discuss it separately, probably on the mailing list instead of on a PR.